### PR TITLE
Partial refactor of versioning_helpers

### PR DESCRIPTION
**Pull request by c4ffein**

### CODE REVIEW

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 1]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
-        FSRemoteManifestNotFound
+    Contains a manifest and his known current, previous and next paths.
+    Unlike ManifestDataAndPaths, attributes are mutable.
+    """
+
+    def __init__(self, manifest, source_path=None, destination_path=None, current_path=None):
+        self.manifest = manifest
+        self.source = source_path
+        self.destination = destination_path
+        self.current = current_path
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: Maybe use `attrs` for this class, so you have a syntax similar to the other data classes.

Also, those would have been a good use case for [dataclasses](https://docs.python.org/3/library/dataclasses.html) if we were using python 3.7.

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 2]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
+                return self._manifest_cache[entry_id][version].manifest
             if timestamp:
                 return next(
                     (
-                        t[2]
-                        for t in manifest_cache[entry_id].values()
-                        if t[0] and t[1] and t[0] <= timestamp <= t[1]
+                        t.manifest
+                        for t in self._manifest_cache[entry_id].values()
+                        if t.early and t.late and t.early <= timestamp <= t.late
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: It looks much better :+1:

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 3]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
-        return versions_list_cache[entry_id]
 
-    async def _get_past_path(entry_id: EntryID, version=None, timestamp=None) -> FsPath:
+class VersionsListCache:
+    """
+    Caches results of the remote_loader.list_versions calls (working on EntryIDs).
+    """
+
+    def __init__(self, remote_loader):
+        self._versions_list_cache = {}
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: How long does this cache live in memory?

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 4]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
-                        t[2]
-                        for t in manifest_cache[entry_id].values()
-                        if t[0] and t[1] and t[0] <= timestamp <= t[1]
+                        t.manifest
+                        for t in self._manifest_cache[entry_id].values()
+                        if t.early and t.late and t.early <= timestamp <= t.late
                     )
                 )
+            return
         except (KeyError, StopIteration):
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: The catching scope is a bit wide. Maybe manage the cache miss and the `StopIteration` separately?

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 5]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
             if timestamp:
                 return next(
                     (
-                        t[2]
-                        for t in manifest_cache[entry_id].values()
-                        if t[0] and t[1] and t[0] <= timestamp <= t[1]
+                        t.manifest
+                        for t in self._manifest_cache[entry_id].values()
+                        if t.early and t.late and t.early <= timestamp <= t.late
                     )
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: Those parenthesis are not necessary.

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 6]
```diff
@@ -50,109 +51,187 @@ class ManifestDataAndPaths(NamedTuple):
...
+        try:
+            name = next(
+                name
+                for name, child_id in parent_manifest.children.items()
+                if child_id == current_id
+            )
+        except StopIteration:
+            raise EntryNotFound(entry_id)
+        else:
+            parts.append(name)
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: This part could be written with a `for-else`, although it's a matter of taste:
```python
# Find the child name
for name, child_id in parent_manifest.children.items():
    if child_id == current_id:
        parts.append(name)
        break
else:
    raise EntryNotFound(entry_id)
```

File: parsec/core/fs/workspacefs/versioning_helpers.py [hunk 7]
```diff
@@ -167,39 +246,39 @@ class ManifestDataAndPaths(NamedTuple):
...
-                except (FSRemoteManifestNotFoundBadVersion, EntryNotFound):
-                    pass
+            async def _populate_source_path(data, entry_id, timestamp):
+                data.source = await _try_get_path_at_timestamp(entry_id, timestamp)
+
+            async def _populate_destination_path(data, entry_id, timestamp):
+                data.destination = await _try_get_path_at_timestamp(entry_id, timestamp)
+
+            async def _populate_current_path(data, entry_id, timestamp):
+                data.current = await _try_get_path_at_timestamp(entry_id, timestamp)
```
Comment (vxgmichel) [inline parsec/core/fs/workspacefs/versioning_helpers.py]: So those attributes are only mutated once?

Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: Yes, I still use this class as a way to pass something that can be modified later by the populate_x_path methods, is it ok or is there a much cleaner way?

File: parsec/core/fs/workspacefs/versioning_helpers.py (hunk 6 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: Changed it to yours as I find it cleaner too

File: parsec/core/fs/workspacefs/versioning_helpers.py (hunk 5 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: Thanks, missed that obvious thing...

File: parsec/core/fs/workspacefs/versioning_helpers.py (hunk 4 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: Done, and I splitted this code to another method to make the main one more readable!

File: parsec/core/fs/workspacefs/versioning_helpers.py (hunk 3 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: It is only needed during the list_version function execution, so I added a comment to explain

File: parsec/core/fs/workspacefs/versioning_helpers.py (hunk 1 above)
Comment (c4ffein) [reply parsec/core/fs/workspacefs/versioning_helpers.py]: Done!

**Pull request merged**